File: sdk-ruby/test/x402_test.rb
# frozen_string_literal: true

# Conformance tests for Wave::X402 (EIP-3009 "exact" scheme signing). Asserts byte-for-byte parity with
# WAVE's reference TypeScript x402 signer via the shared vector (testdata/x402_exact_vector.json) the
# facilitator verifies. Skips cleanly if the optional `eth` gem is not installed (the base SDK and the
# offline smoke suite do not depend on it).

require "minitest/autorun"
require "json"
require_relative "../lib/wave/x402"

begin
  require "eth"
  ETH_AVAILABLE = true
rescue LoadError
  ETH_AVAILABLE = false
end

class X402ConformanceTest < Minitest::Test
  def setup
    skip "eth gem not installed (gem install eth)" unless ETH_AVAILABLE
    path = File.join(__dir__, "..", "..", "testdata", "x402_exact_vector.json")
    @vector = JSON.parse(File.read(path))
  end

  def test_signature_and_header_byte_identical
    @vector["vectors"].each do |v|
      a = v["authorization"]
      payload = Wave::X402.sign_exact_authorization(
        private_key: @vector["payerPrivateKey"],
        network: v["network"],
        to: a["to"],
        value: a["value"],
        valid_before: a["validBefore"],
        valid_after: a["validAfter"],
        nonce: a["nonce"]
      )
      assert_equal v["signature"], payload["signature"], "#{v['network']} signature"
      assert_equal a, payload["authorization"], "#{v['network']} authorization wire shape"
      assert_equal v["header"], Wave::X402.encode_exact_payment_header(v["network"], payload),
                   "#{v['network']} header"
    end
  end

  def test_derives_payer_and_defaults
    a = @vector["vectors"][0]["authorization"]
    payload = Wave::X402.sign_exact_authorization(
      private_key: @vector["payerPrivateKey"], network: "base",
      to: a["to"], value: 1000, valid_before: 1_900_000_600, nonce: a["nonce"]
    )
    assert_equal @vector["payerAddress"].downcase, payload["authorization"]["from"].downcase
    assert_equal "0", payload["authorization"]["validAfter"]
    assert_equal "1000", payload["authorization"]["value"]
  end

  def test_random_nonce_is_bytes32_hex
    n = Wave::X402.random_nonce
    assert n.start_with?("0x")
    assert_equal 66, n.length
    refute_equal Wave::X402.random_nonce, Wave::X402.random_nonce
  end

  def test_guards
    a = @vector["vectors"][0]["authorization"]
    assert_raises(ArgumentError) do
      Wave::X402.sign_exact_authorization(
        private_key: @vector["payerPrivateKey"], network: "base", to: a["to"],
        value: "1", valid_before: "1900000600",
        from: "0x0000000000000000000000000000000000000001"
      )
    end
    assert_raises(ArgumentError) do
      Wave::X402.sign_exact_authorization(
        private_key: @vector["payerPrivateKey"], network: "ethereum", to: a["to"],
        value: "1", valid_before: "1900000600"
      )
    end
  end
end

File: sdk-go/wave/x402_test.go
package wave

// Conformance tests for x402 "exact" signing. Asserts byte-for-byte parity with WAVE's reference
// TypeScript signer via the shared vector (testdata/x402_exact_vector.json) the facilitator verifies.

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

type x402Vector struct {
	PayerPrivateKey string `json:"payerPrivateKey"`
	PayerAddress    string `json:"payerAddress"`
	Vectors         []struct {
		Network       string             `json:"network"`
		Authorization ExactAuthorization `json:"authorization"`
		Signature     string             `json:"signature"`
		Header        string             `json:"header"`
	} `json:"vectors"`
}

func loadX402Vector(t *testing.T) x402Vector {
	t.Helper()
	b, err := os.ReadFile(filepath.Join("..", "..", "testdata", "x402_exact_vector.json"))
	if err != nil {
		t.Fatalf("read vector: %v", err)
	}
	var v x402Vector
	if err := json.Unmarshal(b, &v); err != nil {
		t.Fatalf("parse vector: %v", err)
	}
	return v
}

func TestSignExactAuthorization_ByteIdenticalToReference(t *testing.T) {
	v := loadX402Vector(t)
	for _, vec := range v.Vectors {
		a := vec.Authorization
		payload, err := SignExactAuthorization(SignExactParams{
			PrivateKey:  v.PayerPrivateKey,
			Network:     vec.Network,
			To:          a.To,
			Value:       a.Value,
			ValidBefore: a.ValidBefore,
			ValidAfter:  a.ValidAfter,
			Nonce:       a.Nonce,
		})
		if err != nil {
			t.Fatalf("[%s] sign: %v", vec.Network, err)
		}
		if payload.Signature != vec.Signature {
			t.Errorf("[%s] signature mismatch\n got %s\nwant %s", vec.Network, payload.Signature, vec.Signature)
		}
		if payload.Authorization != a {
			t.Errorf("[%s] authorization wire shape mismatch: %+v vs %+v", vec.Network, payload.Authorization, a)
		}
		hdr, err := EncodeExactPaymentHeader(vec.Network, payload)
		if err != nil {
			t.Fatalf("[%s] header: %v", vec.Network, err)
		}
		if hdr != vec.Header {
			t.Errorf("[%s] header mismatch\n got %s\nwant %s", vec.Network, hdr, vec.Header)
		}
	}
}

func TestSignExactAuthorization_DerivesPayer(t *testing.T) {
	v := loadX402Vector(t)
	payload, err := SignExactAuthorization(SignExactParams{
		PrivateKey: v.PayerPrivateKey, Network: "base",
		To: v.Vectors[0].Authorization.To, Value: "1000", ValidBefore: "1900000600",
		Nonce: v.Vectors[0].Authorization.Nonce,
	})
	if err != nil {
		t.Fatal(err)
	}
	if !strings.EqualFold(payload.Authorization.From, v.PayerAddress) {
		t.Errorf("from: got %s want %s", payload.Authorization.From, v.PayerAddress)
	}
	if payload.Authorization.ValidAfter != "0" {
		t.Errorf("validAfter default: got %s", payload.Authorization.ValidAfter)
	}
}

func TestRandomNonce_Bytes32Hex(t *testing.T) {
	n := RandomNonce()
	if !strings.HasPrefix(n, "0x") || len(n) != 66 {
		t.Fatalf("nonce format: %s", n)
	}
	if RandomNonce() == RandomNonce() {
		t.Error("nonce not fresh per call")
	}
}

func TestSignExactAuthorization_Guards(t *testing.T) {
	v := loadX402Vector(t)
	if _, err := SignExactAuthorization(SignExactParams{
		PrivateKey: v.PayerPrivateKey, Network: "base", To: v.Vectors[0].Authorization.To,
		Value: "1", ValidBefore: "1900000600", From: "0x0000000000000000000000000000000000000001",
	}); err == nil {
		t.Error("expected error on from-mismatch")
	}
	if _, err := SignExactAuthorization(SignExactParams{
		PrivateKey: v.PayerPrivateKey, Network: "ethereum", To: v.Vectors[0].Authorization.To,
		Value: "1", ValidBefore: "1900000600",
	}); err == nil {
		t.Error("expected error on unsupported network")
	}
}
